reposerver/repository/repository.go

```diff
@@ -434,11 +440,55 @@ func (s *Service) runRepoOperation(
 	}
 }
 
+type gitClientGetter func(repo *v1alpha1.Repository, revision string, opts ...git.ClientOpts) (git.Client, string, error)
+
+// resolveReferencedSources resolves the revisions for the given referenced sources. This lets us invalidate the cached
+// when one or more referenced sources change.
+func resolveReferencedSources(hasMultipleSources bool, source *v1alpha1.ApplicationSourceHelm, refSources map[string]*v1alpha1.RefTarget, newClientResolveRevision gitClientGetter) (map[string]string, error) {
+	repoRefs := make(map[string]string)
+	if hasMultipleSources {
+		if source != nil {
+			// Checkout every one of the referenced sources to the target revision before generating Manifests
+			for _, valueFile := range source.ValueFiles {
+				if strings.HasPrefix(valueFile, "$") {
+					refVar := strings.Split(valueFile, "/")[0]
+
+					refSourceMapping, ok := refSources[refVar]
+					if !ok {
+						if len(refSources) == 0 {
+							return nil, fmt.Errorf("source referenced %q, but no source has a 'ref' field defined", refVar)
+						}
+						refKeys := make([]string, 0)
+						for refKey := range refSources {
+							refKeys = append(refKeys, refKey)
+						}
+						return nil, fmt.Errorf("source referenced %q, which is not one of the available sources (%s)", refVar, strings.Join(refKeys, ", "))
+					}
+					if refSourceMapping.Chart != "" {
+						return nil, fmt.Errorf("source has a 'chart' field defined, but Helm charts are not yet not supported for 'ref' sources")
+					}
+					normalizedRepoURL := git.NormalizeGitURL(refSourceMapping.Repo.Repo)
```

Comment:
Should we move this to a new function as we are using the exact same steps [here](https://github.com/argoproj/argo-cd/blob/master/reposerver/repository/repository.go#L623-L642)? WDYT?

Reply:
Per Slack convo, added comments to note the duplication. We can try to clean up in a follow-up refactoring PR after 2.6 is out.